feat(typedoc-appium-plugin): implement cross-referencing of methods

boneskull opened the pull request:
This is getting pretty close, but:

- Commands from a driver will show even w/o a `newMethodMap` or `execMethodMap`; it is able to understand that it's overriding `BaseDriver`
- It's also able to understand the associated routes
- And it's able to understand the routes even if `BaseDriver` does not implement it

~~Still working on the parameters display, but we have the name override working at least.~~

**Update:** Parameters are working fine.

Added some docstrings in `fake-driver`, `base-driver` for testing

boneskull committed:
feat(typedoc-appium-plugin): implement cross-referencing of methods

This is getting pretty close, but:

- Commands from a driver will show even w/o a `newMethodMap` or `execMethodMap`; it is able to understand that it's overriding `BaseDriver`
- It's also able to understand the associated routes
- And it's able to understand the routes even if `BaseDriver` does not implement it

Still working on the parameters display, but we have the name override working at least.

Added some docstrings in `fake-driver`, `base-driver` for testing

boneskull commented:
Current dependencies on/for this PR:
* master
  * **PR #17971** <a href="https://app.graphite.dev/github/pr/appium/appium/17971" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/appium/appium/17971?utm_source=stack-comment).

boneskull commented on packages/typedoc-plugin-appium/lib/converter/comment.ts:
```diff
@@ -0,0 +1,104 @@
+import {Comment, CommentDisplayPart} from 'typedoc';
+import {AsyncMethodDeclarationReflection, KnownMethods} from './types';
+
+export const stats = _.fill(new Array(5), 0);
```
this is for debugging to let me know which "comment finders" are actually being called, so I can remove the ones that don't get called

jlipps reviewed (commented):
just saving my review state

boneskull commented:
@jlipps OK, this is looking pretty good now.  This is what we get from `FakeDriver`:

![image](https://user-images.githubusercontent.com/924465/209237674-a7362504-9174-49b8-b0dd-df0be112d79d.png)

Above we can see that:

1. The comment comes from the `FakeDriver#createSession()` method; it is not in the method map.
2. The parameter names and whether they are required or optional come from the _builtin method map_ in `routes.js`.  
3. The capabilities types come from `FakeDriver` 

Here are some from `FakeDriver`'s method map:

![image](https://user-images.githubusercontent.com/924465/209237895-04d967f4-609e-438e-af0e-eaf47ee7ffc1.png)

1. The `GET` method of `/session/:sessionId/fakedriver` has no description in the new method map _or_ the method itself.
2. The `POST` method of `/session/:sessionId/fakedriver` has a description from the method itself, which takes precedence over the description from the method map (which exists)

And some from the execute method map:

![image](https://user-images.githubusercontent.com/924465/209238115-077fdb82-d969-4ee1-9d06-3a10eaeb4a5a.png)

1. The description of `fake: addition` and `fake: setThing` come from the class methods associated with the execute method
2. The description of `fake: getThing` comes from the execute method map itself, since the class method has no description

boneskull commented:
And there's some output when run with verbose log level, so you can see what's happening:

```
[appium:converter] Converting module @appium/fake-driver
[appium:converter] (FakeDriver) Converting 40 potential methods
[appium:converter] (FakeDriver) Added GET route /session/:sessionId/fakedriver to method getFakeThing for command "getFakeThing"
[appium:converter] (FakeDriver) Added POST route /session/:sessionId/fakedriver to method setFakeThing for command "setFakeThing"
[appium:converter] (FakeDriver) Added GET route /session/:sessionId/fakedriverargs to method getFakeDriverArgs for command "getFakeDriverArgs"
[appium:converter] (FakeDriver) Added /session/:sessionId/execute route to method fakeAddition for command "fakeAddition" from script "fake: addition"
[appium:converter] (FakeDriver) Added /session/:sessionId/execute route to method getFakeThing for command "getFakeThing" from script "fake: getThing"
[appium:converter] (FakeDriver) Added /session/:sessionId/execute route to method setFakeThing for command "setFakeThing" from script "fake: setThing"
[appium:converter] (FakeDriver) Added POST route to method createSession for command "createSession"
[appium:converter] (FakeDriver) Added POST route to method reset for command "reset"
[appium:converter] (FakeDriver) Added GET route to method deleteSession for command "deleteSession"
[appium:converter] (FakeDriver) Added DELETE route to method deleteSession for command "deleteSession"
[appium:converter] (FakeDriver) Added GET route to method getStatus for command "getStatus"
[appium:converter] (FakeDriver) Added GET route to method timeouts for command "timeouts"
[appium:converter] (FakeDriver) Added POST route to method timeouts for command "timeouts"
[appium:converter] (FakeDriver) Added POST route to method implicitWait for command "implicitWait"
[appium:converter] (FakeDriver) Added GET route to method getTimeouts for command "getTimeouts"
[appium:converter] (FakeDriver) Added POST route to method getTimeouts for command "getTimeouts"
[appium:converter] (FakeDriver) Added POST route to method logCustomEvent for command "logCustomEvent"
[appium:converter] (FakeDriver) Added POST route to method getLogEvents for command "getLogEvents"
[appium:converter] (FakeDriver) Added POST route to method findElement for command "findElement"
[appium:converter] (FakeDriver) Added POST route to method findElements for command "findElements"
[appium:converter] (FakeDriver) Added POST route to method findElementFromElement for command "findElementFromElement"
[appium:converter] (FakeDriver) Added POST route to method findElementsFromElement for command "findElementsFromElement"
[appium:converter] (FakeDriver) Added GET route to method getPageSource for command "getPageSource"
[appium:converter] (FakeDriver) Added GET route to method getLogTypes for command "getLogTypes"
[appium:converter] (FakeDriver) Added GET route to method getLogTypes for command "getLogTypes"
[appium:converter] (FakeDriver) Added POST route to method getLog for command "getLog"
[appium:converter] (FakeDriver) Added POST route to method getLog for command "getLog"
[appium:converter] (FakeDriver) Added POST route to method getSettings for command "getSettings"
[appium:converter] (FakeDriver) Added GET route to method getSettings for command "getSettings"
[appium:converter] (FakeDriver) Added GET route to method getSessions for command "getSessions"
[appium:converter] (FakeDriver) Added GET route to method getSession for command "getSession"
[appium:converter] (FakeDriver) Added DELETE route to method getSession for command "getSession"
[appium:converter] (FakeDriver) The following async methods were not found in any method map: updateServer, proxyCommand, executeCommand, startUnexpectedShutdown, startNewCommandTimeout, clearNewCommandTimeout, implicitWaitForCondition, implicitWaitW3C, implicitWaitMJSONWP, pageLoadTimeoutW3C, pageLoadTimeoutMJSONWP, scriptTimeoutW3C, scriptTimeoutMJSONWP, newCommandTimeout, findElOrEls, findElOrElsWithProcessing, executeMethod
[appium:converter] (FakeDriver) Done; found 21 routes and 3 execute methods
```

boneskull commented:
("routes", for statistical purposes above, have a unique name, but each may have 1-3 associated HTTP methods)

boneskull committed:
chore(typedoc-plugin-appium): better sorting, refactors

trying to make some things less opaque.  maybe failed

boneskull committed:
feat: create @appium/tsconfig

For the TypeDoc plugin to correctly pull information from `@appium/types` and `@appium/base-driver`, they both need `tsconfig.json` files.

They currently _have_ `tsconfig.json` files, but they extend base configs in the monorepo root, which is problematic.  Moving the configs into a package seems reasonable.  Further, all packages must publish `tsconfig.json`.  There's precedent for publishing a `tsconfig.json` this way.

FWIW the `paths` and `references` fields in the `tsconfig.json` files seem to be ignored by TypeDoc, so we can get away with this.

boneskull committed:
chore(base-driver): do not use EmptyObject
I do not trust it

jlipps reviewed (approved):
LGTM!
